datafusion/expr/src/binary_rule.rs

```diff
@@ -457,18 +449,18 @@ fn dictionary_coercion(
         (
             DataType::Dictionary(_lhs_index_type, lhs_value_type),
             DataType::Dictionary(_rhs_index_type, rhs_value_type),
-        ) => dictionary_value_coercion(lhs_value_type, rhs_value_type),
```

Comment:
Arrow now supports general purpose casting, so we can just call back into the general comparison coercion code